### PR TITLE
Upgrade elastic-agent-autodiscover dependency; use InitDefaults() to init default configuration

### PR DESCRIPTION
**Pull request by tetianakravchenko**
## Proposed commit message

- WHAT:
  - Upgrade elastic-agent-autodiscover dependency;
  - implement InitDefaults() for `kubeAnnotatorConfig`
  - add missing enriching parameters
- WHY:
  - elastic-agent-autodiscover v0.6.6 includes this [fix](https://github.com/elastic/elastic-agent-autodiscover/pull/72), that was causing unexpected behavior of `add_resource_metadata` configuration.
  - use `InitDefaults` from `go-ucfg` to ensure that the default config is always used before unpacking configuration, similar to https://github.com/elastic/beats/pull/16857
  - add missing enriching parameters to demonstrate that parameters such as `include_labels`, etc can be used on the metricset configuration level

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Relates https://github.com/elastic/elastic-agent/issues/3636

## Use cases

`AddResourceMetadata` is used for 3 different components:
- autodiscover
config:
```
templates:
            - config:
                - module: kubernetes
                  hosts: ["kube-state-metrics:8080"]
                  period: 10s
                  add_resource_metadata:
                    namespace:
                      include_labels: ["labelName"]
                    node:
                      include_labels: ["nodeLabel"]
                    deployment: true
                    cronjob: true
                  metricsets:
                    - state_node
                    - state_deployment
                    - ...
```
Screenshot 2023-12-19 at 10 53 18

- k8s module
config:
```
kubernetes.yml: |-
    - module: kubernetes
      add_resource_metadata:
        namespace:
          include_labels: ["labelName"]
        deployment: true
        cronjob: true
      metricsets:
        - node
        - system
        - pod
        - container
        - volume
```
Screenshot 2023-12-19 at 10 46 03

- processor
filebeat config:
```
filebeat.inputs:
    - type: container
      paths:
        - /var/log/containers/*.log
      processors:
        - add_kubernetes_metadata:
            add_resource_metadata:
              namespace:
                include_labels: ["labelName"]
            host: ${NODE_NAME}
            matchers:
            - logs_path:
                logs_path: "/var/log/containers/"
```
Screenshot 2023-12-19 at 11 01 09

metricbeat - this processor does not work:
- on the top level of the configuration
- on the module level
TODO: update documentation and remove this processor from the metricbeat

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -405,7 +405,7 @@ replace (
 	github.com/docker/go-plugins-helpers => github.com/elastic/go-plugins-helpers v0.0.0-20200207104224-bdf17607b79f
 	github.com/dop251/goja => github.com/andrewkroh/goja v0.0.0-20190128172624-dd2ac4456e20
 	github.com/dop251/goja_nodejs => github.com/dop251/goja_nodejs v0.0.0-20171011081505-adff31b136e6
-
+	github.com/elastic/elastic-agent-autodiscover => /Users/tetianakravchenko/go/src/github.com/elastic/elastic-agent-autodiscover
```
Comment (tetianakravchenko) [inline go.mod:408]: TODO: replace with https://github.com/elastic/elastic-agent-autodiscover/pull/72 when released

Comment (gizas) [reply go.mod:408]: Reminder: We need the same for elastic-agent repo: https://github.com/elastic/elastic-agent/blob/main/go.mod#L15

File: libbeat/processors/add_kubernetes_metadata/config.go [hunk 2]
```diff
@@ -50,15 +50,13 @@ type Enabled struct {
...
-func defaultKubernetesAnnotatorConfig() kubeAnnotatorConfig {
-	return kubeAnnotatorConfig{
-		SyncPeriod:          10 * time.Minute,
-		CleanupTimeout:      60 * time.Second,
-		DefaultMatchers:     Enabled{true},
-		DefaultIndexers:     Enabled{true},
-		Scope:               "node",
-		AddResourceMetadata: metadata.GetDefaultResourceMetadataConfig(),
-	}
+func (c *kubeAnnotatorConfig) InitDefaults() {
```
Comment (tetianakravchenko) [inline libbeat/processors/add_kubernetes_metadata/config.go:53]: refers to the `Unpack` https://github.com/elastic/go-ucfg/blob/main/reify.go#L86-L87

File: metricbeat/module/kubernetes/util/kubernetes.go [hunk 3]
```diff
@@ -161,13 +161,7 @@ func NewResourceMetadataEnricher(
 		return &nilEnricher{}
 	}
 
-	// GetPodMetaGen requires cfg of type Config
-	commonMetaConfig := metadata.Config{}
-	if err := base.Module().UnpackConfig(&commonMetaConfig); err != nil {
```
Comment (tetianakravchenko) [inline metricbeat/module/kubernetes/util/kubernetes.go:166]: I might miss smth here, but I am not sure why it was used here. `cfg` is used later to pass it to [`metadata.GetPodMetaGen`](https://github.com/elastic/beats/pull/37458/files#diff-1d7b17f1ef239e6177db5ee660f2234bfd69c82e93494d4e46ee97f8a8617cfdR192), `metadata.NewServiceMetadataGenerator` and `metadata.NewNamespaceAwareResourceMetadataGenerator`, as result we get the default `metadata.Config`:
Screenshot 2023-12-20 at 15 22 48

All those functions are using `cfg` in `NewResourceMetadataGenerator` - where happens the same: the https://github.com/elastic/elastic-agent-autodiscover/blob/main/kubernetes/metadata/resource.go#L46

**Comment by MichaelKatsoulis** [comment]
@tetianakravchenko can you update the description with what is the motivation of this PR changes. Was it driven from an issue or another PR ?

**Comment by tetianakravchenko** [comment]
@MichaelKatsoulis done, main reason - https://github.com/elastic/elastic-agent/issues/3636

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 4]
```diff
@@ -123,8 +123,7 @@ func New(cfg *config.C) (beat.Processor, error) {
 }
 
 func newProcessorConfig(cfg *config.C, register *Register) (kubeAnnotatorConfig, error) {
-	config := defaultKubernetesAnnotatorConfig()
-
+	var config kubeAnnotatorConfig
```
Comment (MichaelKatsoulis) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go:126]: Upon construction of the config object, the InitDefaults is called under the hood?

Comment (tetianakravchenko) [reply libbeat/processors/add_kubernetes_metadata/kubernetes.go:126]: yes, https://github.com/elastic/go-ucfg/blob/main/initializer.go#L24-L29

[`Unpack`](https://github.com/elastic/go-ucfg/blob/main/reify.go#L137C18-L137C24) calls [reifyInfo](https://github.com/elastic/go-ucfg/blob/main/reify.go#L155), where after detecting the type (Struct) will be called function [`tryInitDefaults`](https://github.com/elastic/go-ucfg/blob/main/reify.go#L261)

It is not a required changed, but I've noticed that it is used in elastic-agent and in beats in multiple places

Comment (tetianakravchenko) [reply libbeat/processors/add_kubernetes_metadata/kubernetes.go:126]: yes, https://github.com/elastic/go-ucfg/blob/main/initializer.go#L24-L29

[`Unpack`](https://github.com/elastic/go-ucfg/blob/main/reify.go#L137C18-L137C24) calls [reifyInfo](https://github.com/elastic/go-ucfg/blob/main/reify.go#L155), where after detecting the type (Struct) will be called function [`tryInitDefaults`](https://github.com/elastic/go-ucfg/blob/main/reify.go#L261)

It is not a required changed, but I've noticed that it is used in elastic-agent and in beats in multiple places

File: metricbeat/module/kubernetes/util/kubernetes.go (hunk 3 above)
Comment (tetianakravchenko) [reply metricbeat/module/kubernetes/util/kubernetes.go:166]: @MichaelKatsoulis as discussed - I've reverted this change and added missing metadata enrichment parameters to the configs https://github.com/elastic/beats/pull/37458/commits/0215971e6ec689b794a4e1c7d970a5ab7c79b01a

File: metricbeat/module/kubernetes/util/kubernetes.go [hunk 5]
```diff
@@ -308,6 +308,14 @@ func NewContainerMetadataEnricher(
...
 	}
 
+	// commonMetaConfig stores the metadata configuration of the resource itself
+	commonMetaConfig := metadata.Config{}
+	if err := base.Module().UnpackConfig(&commonMetaConfig); err != nil {
+		logp.Err("Error initializing Kubernetes metadata enricher: %s", err)
+		return &nilEnricher{}
+	}
+	cfg, _ := conf.NewConfigFrom(&commonMetaConfig)
+
```
Comment (MichaelKatsoulis) [inline metricbeat/module/kubernetes/util/kubernetes.go:318]: Here you just moved the  cfg  creation higher in the code?

**Review by MichaelKatsoulis — APPROVED** [review]

**Review by gizas — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]
Adding approval for files owned by the Elastic Agent team.

**Pull request merged**